[yxtan02] iP

yxtan02 opened the pull request:
# Thanos
>“Snap away your worries, and let Thanos handle the rest.” – Thanos ([source](https://marvelcinematicuniverse.fandom.com/wiki/Thanos))

Thanos frees your mind of having to remember things you need to do. It's:
- Command-line based for simplicity.
- Intuitive and quick to master.
- ~~FAST~~ *SUPER* FAST to use

Here’s how to get started:
1. Download the app's `.jar` file from [here](https://github.com/yxtan02/ip/releases/tag/A-Jar).
2. `Cd` into the folder of the downloaded jar file
3. Run `java -jar thanos.jar`
4. Start adding tasks to your task list! 🔥 
5. Sit back and let Thanos manage your tasks. 😆
 
And it is **COMPLETELY FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ]  GUI (coming soon)

For those who code in Java, here's a look at the app's `main` method:
```java
public class Thanos {
    public static void main(String[] args) {
        new Thanos("data.txt").run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

yxtan02 committed:
Rename chatbot to Thanos and add greeting

yxtan02 committed:
Add echo functionality

yxtan02 committed:
Add input storage and listing

yxtan02 committed:
Add mark and unmark task feature

yxtan02 committed:
Add Todo, Deadline and Event classes

yxtan02 committed:
Set up automated text ui testing and edit chatbot display

yxtan02 committed:
Add error handling

yxtan02 committed:
Add delete task functionality

yxtan02 committed:
Add CommandType enum

yxtan02 committed:
Fix styling

yxtan02 committed:
Add tasks and commands packages

yxtan02 committed:
Add task persistence functionality through file storage

Tasks added by users will be lost after the program is stopped.

Storing these tasks in a local file will allow data to persist.
Let's,
- add a Storage class that can load and save tasks from a file
- add a method in each task class to convert it to a suitable file
string
- add an exception ot handle invalid file formats
- refactor task list into a separate class to reduce code duplication

yxtan02 committed:
Merge branch 'branch-Level-7'

yxtan02 committed:
Move InvalidCommandException to exceptions package

yxtan02 committed:
Use LocalDateTime to represent dates and add find by date command

Dates for tasks were represented as a String

Representing dates as LocalDateTime allows for better input validation
and date/time manipulation and comparison

Let's,
- Add DateTimeUtility class to parse string to LocalDateTime
- Change data type for dates to LocalDateTime
- Add date command to find tasks by date

yxtan02 committed:
Merge branch 'branch-Level-8'

yxtan02 committed:
Add UI and various command classes

yxtan02 committed:
Add parser class

yxtan02 committed:
Refactor code to use function from UI class

yxtan02 committed:
Merge branch 'branch-A-MoreOOP'

yxtan02 committed:
Move all files into thanos package

yxtan02 committed:
Merge branch 'branch-A-Packages'

yxtan02 committed:
Merge branch 'add-gradle-support'

yxtan02 committed:
Configure gradle to use checkstyle

yxtan02 committed:
Add JUnit tests

yxtan02 committed:
Merge branch 'branch-A-JUnit'

yxtan02 committed:
Package app as a JAR file

yxtan02 committed:
Merge branch 'branch-A-Jar'

yxtan02 committed:
Add JavaDoc comments

yxtan02 committed:
Merge branch 'branch-A-JavaDoc'

yxtan02 committed:
Fix style errors

yxtan02 committed:
Add find functionality for users to search by keyword

yxtan02 committed:
Add tests for find functionality

yxtan02 committed:
Merge branch 'branch-A-CodingStandard'

yxtan02 committed:
Merge branch 'branch-Level-9'

- Resolve merge conflict for src/main/java/thanos/tasks/TaskList.java

yxtan02 committed:
Add JavaDocs for find task methods

yxtan02 committed:
Edit comment

yxtan02 committed:
Merge branch 'branch-A-CheckStyle'

yxtan02 committed:
Refactor Ui class into ResponseFormatter class

yxtan02 committed:
Add InvalidDateException class

yxtan02 committed:
Add GUI using JavaFX

yxtan02 committed:
Merge branch 'branch-Level-10'

yxtan02 committed:
Use varargs as function parameter

yxtan02 committed:
Merge branch 'branch-A-Varargs'

Jaynon commented on src/main/java/thanos/Launcher.java:
```diff
@@ -0,0 +1,29 @@
+package thanos;
+
+import javafx.application.Application;
+import thanos.ui.Main;
```
It would be better if imports from different packages are separated.
```suggestion
import javafx.application.Application;

import thanos.ui.Main;
```

Jaynon commented on src/main/java/thanos/commands/Command.java:
Nice job leaving very detailed javadoc descriptions :)

Jaynon commented on src/main/java/thanos/commands/CommandType.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    public static CommandType getCommandType(String command) {
+        switch (command) {
+        case "bye":
+            return BYE;
+        case "date":
+            return DATE;
+        case "deadline":
+            return DEADLINE;
+        case "delete":
+            return DELETE;
+        case "event":
+            return EVENT;
+        case "find":
+            return FIND;
+        case "list":
+            return LIST;
+        case "mark":
+            return MARK;
+        case "todo":
+            return TODO;
+        case "unmark":
+            return UNMARK;
+        default:
+            return INVALID;
```
It is good that switch-case statements are in the correct form without indentations

Jaynon commented on src/main/java/thanos/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,62 @@
+    /**
+     * Executes the command to delete a task from the {@code TaskList}.
+     * <p>
+     * This method parses the command argument to obtain the task index, verifies the format and index validity,
+     * removes the task from the {@code TaskList}, and returns a formatted string confirming the removal of the task
+     * and displaying the updated task count. If the argument is empty, incorrectly formatted, or if the index is
+     * invalid, an {@code InvalidCommandException} is thrown.
+     * </p>
```
It would be nicer if the description would be less verbose and the details be placed as comments within the code itself, so that it will be easy for others to follow.

Jaynon reviewed (commented):
Very clean code and detailed comments! Very minor styling issues and verbose comments which can be shortened and placed as comments in the code itself.

minhao23 commented on src/main/java/thanos/commands/CommandType.java:
```diff
@@ -0,0 +1,56 @@
+ * Enum representing the various types of commands that can be executed within the application.
+ */
+public enum CommandType {
+    BYE,
```
great to see you using enums to streamline your code!

minhao23 commented on src/main/java/thanos/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,62 @@
+ * Represents a command to delete a task from the {@code TaskList}.
+ * It expects an index of the task to be deleted, which is used to locate and remove the task from the list.
+ */
+public class DeleteCommand extends Command {
```
regarding naming, good to see you use nouns for the command classes

minhao23 commented on src/main/java/thanos/tasks/TaskList.java:
```diff
@@ -0,0 +1,141 @@
+package thanos.tasks;
```
this entire file uses apt verbs for its methods, good job!

minhao23 commented on src/main/java/thanos/utility/ResponseFormatter.java:
```diff
@@ -0,0 +1,49 @@
+     * @param task The task that was added.
+     * @param size The current number of tasks in the list.
+     * @return A formatted string indicating the task has been added and showing the current size of the task list.
+     */
```
great use of naming

minhao23 commented on src/test/java/thanos/tasks/TaskListTest.java:
```diff
@@ -0,0 +1,92 @@
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+
+import org.junit.jupiter.api.BeforeEach;
```
impressive that you have a test for every class!

minhao23 reviewed (commented):
LGTM! great job!

yxtan02 committed:
Enable assertions

yxtan02 committed:
Add assertions

The code assumes that the file name provided to Storage is not empty
and that the Storage object provided to TaskList is not null.

This can lead to exceptions being thrown, which are hard to debug.

Add assertions to verify that such assumptions are true and to provide
informative error messages if such assumptions are false.

yxtan02 committed:
Improve code quality

Some parts of the code did not abide by the SLAP principle or were too
deeply nested. This undermines the readability of the code.

Let's refactor the code to address such coding issues.

yxtan02 committed:
Refactor TaskList to use streams

Refactor task filtering logic in TaskList to utilize Java Streams. This
is to improve code readability and conciseness.

yxtan02 committed:
Merge pull request #1 from yxtan02/branch-A-Assertions

Add assertions

yxtan02 committed:
Merge branch 'master' into branch-A-CodeQuality

yxtan02 committed:
Merge branch 'master' into branch-A-Streams

yxtan02 committed:
Merge pull request #2 from yxtan02/branch-A-CodeQuality

Improve code quality

yxtan02 committed:
Merge branch 'master' into branch-A-Streams

yxtan02 committed:
Merge pull request #3 from yxtan02/branch-A-Streams

Refactor TaskList to use streams

yxtan02 committed:
Add gradle.yml to set up GitHub Actions

yxtan02 committed:
Edit junit tests and fix bug in TaskList

yxtan02 committed:
Merge pull request #4 from yxtan02/branch-A-CI

Add gradle.yml to set up GitHub Actions

yxtan02 committed:
Add mass delete functionality

yxtan02 committed:
Update junit test for DeleteCommand

yxtan02 committed:
Merge pull request #5 from yxtan02/branch-BCD-Extension

Add MassOps

yxtan02 committed:
Improve GUI

yxtan02 committed:
Merge pull request #6 from yxtan02/branch-A-BetterGui

Improve GUI

yxtan02 committed:
Add error handling for event start/end times

yxtan02 committed:
Edit error messages to be more concise

yxtan02 committed:
Merge pull request #7 from yxtan02/branch-A-MoreErrorHandling

Add more error handling

yxtan02 committed:
Add more junit tests

yxtan02 committed:
Merge pull request #8 from yxtan02/branch-A-MoreTesting

Add more junit tests

yxtan02 committed:
Clean up Javadocs

yxtan02 committed:
Add chatbot name to GUI

yxtan02 committed:
Add screenshot of chatbot

yxtan02 committed:
Update README.md

yxtan02 committed:
Update screenshot

yxtan02 committed:
Merge branch 'branch-A-UserGuide'

yxtan02 committed:
Refactor code to remove magic literals

yxtan02 committed:
Update README.md

yxtan02 committed:
Update README.md

yxtan02 committed:
Update README.md

yxtan02 committed:
Refactor code to follow SLAP

yxtan02 committed:
Update README.md

yxtan02 committed:
Fix alignment issue for GUI
